src/Microsoft.IdentityModel.Tokens/EventBasedLRUCache.cs

```diff
@@ -0,0 +1,344 @@
+        private readonly double _maxCapacityPercentage = .95;
+        private bool _disposed = false;
+        private readonly int _tryTakeTimeout;
+        // if true, expired values will not be added to the cache and clean-up of expired values will occur on a 5 minute interval
+        private readonly bool _removeExpiredValues;
+
+        internal EventBasedLRUCache(
+            int capacity,
+            TaskCreationOptions options = TaskCreationOptions.LongRunning,
+            IEqualityComparer<TKey> comparer = null,
+            int tryTakeTimeout = 500,
+            bool removeExpiredValues = false,
+            int cleanUpIntervalInSeconds = 300)
+        {
+            _tryTakeTimeout = tryTakeTimeout;
+            _capacity = capacity > 0 ? capacity : throw LogHelper.LogExceptionMessage(new ArgumentOutOfRangeException(nameof(capacity)));
+            _map = new ConcurrentDictionary<TKey, LRUCacheItem<TKey, TValue>>(comparer ?? EqualityComparer<TKey>.Default);
+            _removeExpiredValues = removeExpiredValues;
+            _eventQueueTask = new Task(OnStart, options);
+            _eventQueueTask.Start();
+            if (_removeExpiredValues)
+                _ = RemoveExpiredValuesPeriodically(TimeSpan.FromSeconds(cleanUpIntervalInSeconds));
+        }
+
+        private void OnStart()
+        {
+            while (!_disposed)
+            {
+                try
+                {
+                    if (_eventQueue.TryTake(out var action, _tryTakeTimeout))
+                        action.Invoke();
+                }
+                catch (Exception ex)
+                {
+                    LogHelper.LogWarning(LogHelper.FormatInvariant(LogMessages.IDX10900, ex));
+                }          
+            }
+        }
+
+        public bool Contains(TKey key)
+        {
+            if (key == null)
+                throw LogHelper.LogArgumentNullException(nameof(key));
+
+            return _map.ContainsKey(key);
+        }
+
+        /// <summary>
+        /// FOR TESTING PURPOSES ONLY.
+        /// </summary>
+        internal void WaitForProcessing()
+        {
+            while (true)
+            {
+                if (_eventQueue.Count == 0)
+                    return;
+            }
+        }
+
+        internal int RemoveExpiredValues()
+        {
+            int numItemsRemoved = 0;
+            var node = _doubleLinkedList.First;
+            while (node != null)
+            {
+                var nextNode = node.Next;
+                if (node.Value.ExpirationTime < DateTime.UtcNow)
+                {
+                    _doubleLinkedList.Remove(node);
+                    _map.TryRemove(node.Value.Key, out _);
+                    numItemsRemoved++;
+                }
+
+                node = nextNode;
+            }
+
+            return numItemsRemoved;
+        }
+
+        internal void RemoveLRUs()
```

Comment:
>RemoveLRUs [](start = 22, length = 10)

this might need to check for isdisposed? (not sure about try/catch)

Reply:
This will run in the context of in OnStart() which will be wrapped in a try catch